docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+            if me.X() = pOther.X() and me.Y() = pOther.Y() then
+                return true
+            else
+                return false
+            end if
+        end method
+
+        method FormatAsString(in me) as string
+            return "(" & me.X() formatted as string & ", " & \
+                   me.Y() formatted as string & ")"
+        end method
+    end trait
+
+Traits can be derived from multiple supertraits.
+
+## Trait objects
+
+    public trait U
+        method type F(in me)
+    end trait
+
+    type T
+
+    implementation of U for T
+        method F(in me)
+        end method
+    end implementation
+
+    public handler G(in pParam as U)
+        pParam.F()
+    end handler
+
+A trait object is created whenever a concrete type `T` needs to be passed by
+trait `U` rather than directly as an `T`.  It packages the `T` value with
+information that identifies the methods of the implementation of `U` that was
+used.
+
+## Further work
+
+### Parameterized types
+
+Traits do not fully solve the problem of implementing `formatted as string`
+syntax with full generality and type safety.  There remains the problem of
+formatting polymorphic compound types such as `list`.  Consider the following
+naïve approach:
+
+    public implementation of CanFormatAsString for list
+        method FormatAsString(in me) as string
+            variable tString as string
+            put "(" into tString
+
+            variable tIter as any
+            repeat for each element tElement of me
+                variable tElement as CanFormatAsString
+                put tIter into tElement -- TYPE UNSAFE
+                put tElement formatted as string after tString
+            end repeat
+
+            put ")" after tString
+            return tString
+        end method
+    end implementation
+
+If the above method is compiled with strict type checking, then the marked line
+should fail to compile because `tIter` is of unknown type `any`, and therefore
+does not necessarily implement the `CanFormatAsString` trait.
+
+It would be better if the `CanFormatAsString` trait was only defined for lists
+where every element also conforms to `CanFormatAsString`.
+
+Similar arguments apply for the `array` type.
+
+To get full generality and type safety of the `formatted as string` trait
+Builder therefore also needs type specialization (e.g. `list of
+CanFormatAsString`).
+
+### Interaction with runtime typing
+
+    trait type T
+    end trait
+
+    implementation of T for integer
+    end implementation
+
+    handler Confuse(in pIn as T) as T
+        return pIn
+    end handler
+
+    public handler main()
+        variable tVar as any
+        put 15 into tVar        -- contents of tVar is an integer
+        Confuse(tVar) into tVar -- contents of tVar is a T trait object
+        add 1 to tVar           -- error?
+    end handler
+
+It is not clear what the correct behaviour is here.
```

Comment:
What the 'correct' behavior here does depend on how we view a trait T.

One view (distinct) would be that implementing a trait T for type U creates a new type (T, U) where (T, U) conforms to any other type (T, V) and U bridges to (T, U). With this point of view the above would give an error (unless 'add' had an overload for T). This would require some sort of indirection to implement (e.g. a trait object) which means that an implicit (bridging) operation is needed when going from U to (T, U).

Another view (interface) would be that implementing a trait T for type U makes U conform to T. WIth this point of view the above would not give an error as the runtime type of tVar would remain 'integer'. This would mean that the method invocation operator '.' would actually be a typed operation 'call method on value for trait T'.

Advantages of the 'distinct' view is that it gives a stronger interface guarantee - however, the presence of an implicit operation which is based on a type means that you would get different behavior if you removed type annotations from slots (inserting explicit 'as' operators as needed).

Advantages of the 'interface' view is that you can remove type annotations from slots (inserting explicit 'as' operators as needed) and you will still get the same behavior - however this is at the potential expense of subtle binary interface changes. (Java has a similar issue with the presence of interfaces - subclasses can have extra interface implementations, so the set of interfaces implemented by an object returned by a method can change even though the signature of the method does not).